pkg/csi/service/wcpguest/controller.go

```diff
@@ -1281,16 +1295,186 @@ func (c *controller) CreateSnapshot(ctx context.Context, req *csi.CreateSnapshot
 	*csi.CreateSnapshotResponse, error) {
 	ctx = logger.NewContextWithLogger(ctx)
 	log := logger.GetLogger(ctx)
+	start := time.Now()
+	volumeType := prometheus.PrometheusBlockVolumeType
 	log.Infof("CreateSnapshot: called with args %+v", *req)
-	return nil, status.Error(codes.Unimplemented, "")
+	isBlockVolumeSnapshotWCPEnabled := commonco.ContainerOrchestratorUtility.IsFSSEnabled(ctx,
+		common.BlockVolumeSnapshot)
+	if !isBlockVolumeSnapshotWCPEnabled {
+		return nil, logger.LogNewErrorCode(log, codes.Unimplemented, "createSnapshot")
+	}
+	createSnapshotInternal := func() (*csi.CreateSnapshotResponse, error) {
+		// Search for supervisor PVC and ensure it exists
+		supervisorPVCName := req.SourceVolumeId
+		log.Infof("Checking if supervisor PVC %s/%s exists..", c.supervisorNamespace, supervisorPVCName)
+		_, err := c.supervisorClient.CoreV1().PersistentVolumeClaims(c.supervisorNamespace).Get(
+			ctx, supervisorPVCName, metav1.GetOptions{})
+		if err != nil {
+			if errors.IsNotFound(err) {
+				log.Errorf("the supervisor PVC: %s/%s was not found while attempting to take snapshot",
+					c.supervisorNamespace, supervisorPVCName)
+			}
+			return nil, err
+		}
+		// Get supervisor VolumeSnapshotClass.
+		var supervisorVolumeSnapshotClass string
+		for param := range req.Parameters {
+			paramName := strings.ToLower(param)
+			if paramName == common.AttributeSupervisorVolumeSnapshotClass {
+				supervisorVolumeSnapshotClass = req.Parameters[param]
+			}
+		}
+		// Generate the supervisor VolumeSnapshot name
+		// Assuming snapshot prefix is "snapshot-"
+		supervisorVolumeSnapshotName := c.tanzukubernetesClusterUID + "-" + req.Name[9:]
```

Comment:
When you say "snapshot prefix is snapshot-", what snapshot are you referring to?  Do you refer to the supervisorVolumeSnapshotName or the input snapshot name from CreateSnapshotRequest?  On line 1329, supervisorVolumeSnapshotName has a prefix of clusterUID which is not "snapshot".  Do you mean is it the beginning of req.Name?  Please clarify.

Reply:
The prefix `snapshsot-` is associated with the VolumeSnapshotContent created on the guest(due to VolumeSnapshot creation on the guest). This is passed in the `CreateSnapshot` request from the `external-snapshotter`. It's the VolumeSnapshotContent uuid prefixed with `snapshot-`.

`CreateSnapshotRequest.Name` is not the same as `VolumeSnapshot.metadata.Name`.

For Example:

```
apiVersion: snapshot.storage.k8s.io/v1
kind: VolumeSnapshot
metadata:
  annotations:
  creationTimestamp: '2022-03-18T23:00:51Z'
  finalizers:
    - snapshot.storage.kubernetes.io/volumesnapshot-as-source-protection
    - snapshot.storage.kubernetes.io/volumesnapshot-bound-protection
  generation: 1
  managedFields:
  name: example-vanilla-block-snapshot-size-1
  namespace: default
  resourceVersion: '10042'
  uid: bd5bf716-a9a8-4ce6-aa02-1720b02b3f18
  selfLink: >-
    /apis/snapshot.storage.k8s.io/v1/namespaces/default/volumesnapshots/example-vanilla-block-snapshot-size-1
status:
  boundVolumeSnapshotContentName: snapcontent-bd5bf716-a9a8-4ce6-aa02-1720b02b3f18
  creationTime: '2022-03-18T23:00:52Z'
  readyToUse: true
  restoreSize: 1Gi
spec:
  source:
    persistentVolumeClaimName: example-vanilla-block-pvc
  volumeSnapshotClassName: example-vanilla-block-snapclass


apiVersion: snapshot.storage.k8s.io/v1
kind: VolumeSnapshotContent
metadata:
  creationTimestamp: '2022-03-18T23:00:51Z'
  finalizers:
    - snapshot.storage.kubernetes.io/volumesnapshotcontent-bound-protection
  generation: 1
  managedFields:
  name: snapcontent-bd5bf716-a9a8-4ce6-aa02-1720b02b3f18
  resourceVersion: '10041'
  uid: 8dbbcae3-3766-4b67-bbf5-12a598588166
  selfLink: >-
    /apis/snapshot.storage.k8s.io/v1/volumesnapshotcontents/snapcontent-bd5bf716-a9a8-4ce6-aa02-1720b02b3f18
status:
  creationTime: 1647644452371000000
  readyToUse: true
  restoreSize: 1073741824
  snapshotHandle: 527b55dc-c7db-4574-9226-2e33318b06a3+646adf7b-e031-434d-8c99-ad945f7f5182
spec:
  deletionPolicy: Delete
  driver: csi.vsphere.vmware.com
  source:
    volumeHandle: 527b55dc-c7db-4574-9226-2e33318b06a3
  volumeSnapshotClassName: example-vanilla-block-snapclass
  volumeSnapshotRef:
    apiVersion: snapshot.storage.k8s.io/v1
    kind: VolumeSnapshot
    name: example-vanilla-block-snapshot-size-1
    namespace: default
    resourceVersion: '10008'
    uid: bd5bf716-a9a8-4ce6-aa02-1720b02b3f18
```
The above VolumeSnapshotContent object triggers a `CreateSnapshot` with `CreateSnapshotRequest.Name` as `snapshot-8dbbcae3-3766-4b67-bbf5-12a598588166`, which is "snapshot-" + VolumeSnapshotContent.metadata.uid

To summarize, we are trying to create the name of the `VolumeSnapshot` CR that needs to be created on the supervisor, i.e, `VolumeSnapshot.metadata.Name`.

I hope I haven't confused you further.